Crash in zls/src/semantic_tokens.zig:185:45; `const text = self.handle.tree.source[self.previous_position..start];`

[Fuzzing group for this issue ↗](https://fuzzing.zigtools.org/group/504baeab186badc2b346c2737a82a894) (504baeab186badc2b346c2737a82a894)

The following pieces of information may be outdated; please go to the fuzzing webpage to obtain newer samples if that is the case.

## Generic error output

```log
panic: start index 45 is larger than end index 0
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:185:45: 0x411e8d in addDirect (zls)
        const text = self.handle.tree.source[self.previous_position..start];
                                            ^
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:180:41: 0x4131fe in handleComments (zls)
            try self.addDirect(TokenType.comment, mods, comment_start, length);
                                        ^
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:90:54: 0x499a16 in add (zls)
        try self.handleComments(if (token > 0) starts[token - 1] else 0, next_start);
                                                     ^
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:209:24: 0x41f457 in writeNodeTokens (zls)
        try builder.add(ti, tok_type, tok_mod);
                       ^
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:279:9: 0x49b401 in callWriteNodeTokens__anon_35191 (zls)
        return @call(.auto, writeNodeTokens, args);
        ^
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:1006:75: 0x49afaa in writeContainerField (zls)
        try callWriteNodeTokens(allocator, .{ builder, container_field.ast.type_expr });
                                                                          ^
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:304:35: 0x41827c in writeNodeTokens (zls)
        => try writeContainerField(builder, node, .field),
                                  ^
/home/runner/work/zls/zls/zls/src/semantic_tokens.zig:1038:24: 0x413d11 in writeAllSemanticTokens (zls)
        writeNodeTokens(&amp;builder, child) catch |err| switch (err) {
                       ^
/home/runner/work/zls/zls/zls/src/Server.zig:2104:120: 0x3d5019 in semanticTokensFullHandler (zls)
    const token_array = try semantic_tokens.writeAllSemanticTokens(server.arena, &amp;server.document_store, handle, server.offset_encoding);
                                                                                                                       ^
/home/runner/work/zls/zls/zls/src/Server.zig:3139:37: 0x38f598 in processMessage (zls)
            const response = handler(server, params) catch |err| {
                                    ^
/home/runner/work/zls/zls/zls/src/Server.zig:3017:26: 0x33fbf3 in processJsonRpc (zls)
    server.processMessage(message) catch |err| switch (message) {
                         ^
/home/runner/work/zls/zls/zls/src/main.zig:89:30: 0x33eb89 in loop (zls)
        server.processJsonRpc(&amp;arena, json_message);
                             ^
/home/runner/work/zls/zls/zls/src/main.zig:402:13: 0x34078b in main (zls)
    try loop(&amp;server, record_file, replay_file);
            ^
/opt/hostedtoolcache/zig/zig-linux-x86_64-0.11.0-dev.1413+a51c76541/x64/lib/std/start.zig:616:37: 0x329379 in posixCallMainAndExit (zls)
            const result = root.main() catch |err| {
                                    ^
/opt/hostedtoolcache/zig/zig-linux-x86_64-0.11.0-dev.1413+a51c76541/x64/lib/std/start.zig:376:5: 0x328de1 in _start (zls)
    @call(.never_inline, posixCallMainAndExit, .{});
    ^

```

## Latest (at time of filling) entry

|  | Zig | ZLS
| --- | --- | --- |
| Versions | 0.11.0-dev.1420+3cb1ab0e0 | 0.11.0-dev.142+ed9cf06 |
| Files | [Error ↗](https://fuzzing.zigtools.org/entry/1674524792370/stderr.log) | [Input ↗](https://fuzzing.zigtools.org/entry/1674524792370/stdin.log) |
|  | [Output ↗](https://fuzzing.zigtools.org/entry/1674524792370/stdout.log) | [Zig Repro ↗](https://fuzzing.zigtools.org/entry/1674524792370/principal.zig) |

--- a/src/ast.zig
+++ b/src/ast.zig
@@ -512,7 +512,9 @@ pub fn lastToken(tree: Ast, node: Ast.Node.Index) Ast.TokenIndex {
                 n = tree.extra_data[cases.end - 1]; // last case
             }
         },
-        .container_decl_arg => {
+        .container_decl_arg,
+        .container_decl_arg_trailing,
+         => {
             const members = tree.extraData(datas[n].rhs, Node.SubRange);
             if (members.end - members.start == 0) {
                 end_offset += 3; // for the rparen + lbrace + rbrace
@@ -536,7 +538,6 @@ pub fn lastToken(tree: Ast, node: Ast.Node.Index) Ast.TokenIndex {
         },
         .array_init_comma,
         .struct_init_comma,
-        .container_decl_arg_trailing,
         .switch_comma,
         => {
             if (datas[n].rhs != 0) {

--- a/src/ast.zig
+++ b/src/ast.zig
@@ -254,29 +254,29 @@ pub fn forFull(tree: Ast, node: Node.Index) full.While {
 
 pub fn fullPtrType(tree: Ast, node: Node.Index) ?full.PtrType {
     return switch (tree.nodes.items(.tag)[node]) {
-        .ptr_type_aligned => tree.ptrTypeAligned(node),
-        .ptr_type_sentinel => tree.ptrTypeSentinel(node),
-        .ptr_type => tree.ptrType(node),
-        .ptr_type_bit_range => tree.ptrTypeBitRange(node),
+        .ptr_type_aligned => ptrTypeAligned(tree, node),
+        .ptr_type_sentinel => ptrTypeSentinel(tree, node),
+        .ptr_type => ptrTypeSimple(tree, node),
+        .ptr_type_bit_range => ptrTypeBitRange(tree, node),
         else => null,
     };
 }
 
 pub fn fullIf(tree: Ast, node: Node.Index) ?full.If {
     return switch (tree.nodes.items(.tag)[node]) {
-        .if_simple => tree.ifSimple(node),
-        .@"if" => tree.ifFull(node),
+        .if_simple => ifSimple(tree, node),
+        .@"if" => ifFull(tree, node),
         else => null,
     };
 }
 
 pub fn fullWhile(tree: Ast, node: Node.Index) ?full.While {
     return switch (tree.nodes.items(.tag)[node]) {
-        .while_simple => tree.whileSimple(node),
-        .while_cont => tree.whileCont(node),
-        .@"while" => tree.whileFull(node),
-        .for_simple => tree.forSimple(node),
-        .@"for" => tree.forFull(node),
+        .while_simple => whileSimple(tree, node),
+        .while_cont => whileCont(tree, node),
+        .@"while" => whileFull(tree, node),
+        .for_simple => forSimple(tree, node),
+        .@"for" => forFull(tree, node),
         else => null,
     };
 }

--- a/src/semantic_tokens.zig
+++ b/src/semantic_tokens.zig
@@ -74,9 +74,7 @@ const Builder = struct {
         const starts = tree.tokens.items(.start);
         const next_start = starts[token];
 
-        if (next_start < self.previous_position) {
-            return error.MovedBackwards;
-        }
+        if (next_start < self.previous_position) return;
 
         if (self.previous_token) |prev| {
             // Highlight gaps between AST nodes. These can contain comments or malformed code.
@@ -182,6 +180,8 @@ const Builder = struct {
     }
 
     fn addDirect(self: *Builder, tok_type: TokenType, tok_mod: TokenModifiers, start: usize, length: usize) !void {
+        if (start < self.previous_position) return;
+
         const text = self.handle.tree.source[self.previous_position..start];
         const delta = offsets.indexToPosition(text, text.len, self.encoding);
 
@@ -261,13 +261,8 @@ fn colorIdentifierBasedOnType(builder: *Builder, type_node: analysis.TypeWithHan
     }
 }
 
-const WriteTokensError = error{
-    OutOfMemory,
-    MovedBackwards,
-};
-
 /// HACK self-hosted has not implemented async yet
-fn callWriteNodeTokens(allocator: std.mem.Allocator, args: anytype) WriteTokensError!void {
+fn callWriteNodeTokens(allocator: std.mem.Allocator, args: anytype) error{OutOfMemory}!void {
     if (zig_builtin.zig_backend == .other or zig_builtin.zig_backend == .stage1) {
         const FrameSize = @sizeOf(@Frame(writeNodeTokens));
         var child_frame = try allocator.alignedAlloc(u8, std.Target.stack_align, FrameSize);
@@ -280,7 +275,7 @@ fn callWriteNodeTokens(allocator: std.mem.Allocator, args: anytype) WriteTokensE
     }
 }
 
-fn writeNodeTokens(builder: *Builder, maybe_node: ?Ast.Node.Index) WriteTokensError!void {
+fn writeNodeTokens(builder: *Builder, maybe_node: ?Ast.Node.Index) error{OutOfMemory}!void {
     const node = maybe_node orelse return;
 
     const handle = builder.handle;
@@ -1035,10 +1030,7 @@ pub fn writeAllSemanticTokens(
     // reverse the ast from the root declarations
     var buf: [2]Ast.Node.Index = undefined;
     for (ast.declMembers(handle.tree, 0, &buf)) |child| {
-        writeNodeTokens(&builder, child) catch |err| switch (err) {
-            error.MovedBackwards => break,
-            else => |e| return e,
-        };
+        try writeNodeTokens(&builder, child);
     }
     try builder.finish();
     return builder.toOwnedSlice();

--- a/tests/lsp_features/semantic_tokens.zig
+++ b/tests/lsp_features/semantic_tokens.zig
@@ -32,6 +32,18 @@ test "semantic tokens - comments" {
     // TODO more tests
 }
 
+test "semantic tokens - string literals" {
+    // https://github.com/zigtools/zls/issues/921
+    try testSemanticTokens(
+        \\"
+        \\"",// 
+        \\"": 
+    ,
+        // no idea if this output is correct but at least it doesn't crash
+        &.{ 1, 3, 3, 8, 0, 1, 0, 2, 4, 0, 0, 0, 2, 9, 0 },
+    );
+}
+
 const file_uri = switch (builtin.os.tag) {
     .windows => "file:///C:/test.zig",
     else => "file:///test.zig",

